### ISSUE TITLE
Upsert content types and fields

### ISSUE THREAD

**Issue by jody-zeitler** [contentful/migration-cli#21]
I'd like to be able to update spaces that may be partially migrated, such that some content types and fields may already exist in the target space in one form or another. However the create and edit methods exposed by the API are exclusive and validation will fail if the space is in the wrong state. It would be nice to have "upsertContentType" and "upsertField" methods that continue the migration regardless of the whether the type/field already exists.

I tried rolling my own upsert functions with try/catch but that obviously doesn't work since the validator doesn't throw. Is there any way to accomplish this with the current API?

```js
module.exports = {
	upsertContentType(migration, contentTypeId) {
		try {
			return migration.createContentType(contentTypeId);
		} catch (error) {
			return migration.editContentType(contentTypeId);
		}
	},

	upsertField(contentType, fieldId) {
		try {
			return contentType.createField(fieldId);
		} catch (error) {
			return contentType.editField(fieldId);
		}
	},
}
```

**Comment by zcei** [comment]
Hey @jody-zeitler 

you're right that the validator doesn't throw, as this would result in a very annoying "one error at a time" feedback loop. We're trying to provide you with as much information about what is going on with your migration at once. That also means we don't execute every single step "as you write them", but chunk things up (e.g. we can't create a Content Type while you write `createContentType` as required properties like `name` are missing)

What is possible (but not documented before the release of the migration standalone lib) is to return a promise in your migration script. (This also enables you to use `async/await` in case you're on a Node version that supports it)
With this information you could build it yourself by fetching the Content Type in your migration and handle the existence of specific fields. Or catch the error if no Content Type exists).

@TimBeyer suggested something like this:

```js
return sdk.getContentType('foo').then((ct) => {
  migration.editContentType('foo').createField('bar')
}).catch(() => {
  migration.createContentType('foo')
});
```

Is this sufficient for your use-case? 

Best,
Stephan

**Comment by jody-zeitler** [comment]
Is `sdk` the `contentful-management` module? I see `getContentType` exists on Space instances. Is there any way to access the client that the migration is using so I don't have to sort out the spaceId and accessToken myself?

**Comment by zcei** [comment]
Ah true, this was an oversimplified solution. `sdk` should reference a space here, correct. So something like this first: 

```js
contentfulManagement.createClient(yourConfig).getSpace(yourSpace)
```

This is rather a workaround than longterm. I could potentially see passing in `makeRequest` to the migration function as a second parameter for such cases, or a `config` object with e.g. `spaceId` - **this isn't decided yet!**

Especially because it uses `rawRequest` directly rather than the convenience functions this would yield a difference experience than you're used to.

https://github.com/contentful/migration-cli/blob/7f50a16ea66f25cdecd86f241575a3fc302d8154/bin/contentful-migration.js#L80-L84

```js
// with management SDK
space.getContentType('dog')
// all the convenience functions like contentType.publish()

// with makeRequest
request({
  method: 'GET',
  url: '/content_types/dog',
});
// only JSON object response
```

**Comment by jody-zeitler** [comment]
I've got something that _should_ work, but I'm seeing some odd behavior when resolving Promises with migration ContentTypes.

util.js:
```js
const {createClient} = require('contentful-management');
const yargs = require('yargs');

const argv = yargs
	.option('space-id', {alias: 's'})
	.option('access-token', {alias: 'a'})
	.demandOption(['space-id', 'access-token'], 'Provide a space ID and access token.')
	.argv;
const config = {
	accessToken: argv.accessToken,
	spaceId: argv.spaceId
};
const client = createClient(config);

async function getContentType(contentTypeId) {
	const space = await client.getSpace(config.spaceId);
	return space.getContentType(contentTypeId);
}

module.exports = {
	async upsertContentType(migration, contentTypeId) {
		try {
			const ct = await getContentType(contentTypeId);
			return migration.editContentType(contentTypeId);
		} catch (error) {
			return migration.createContentType(contentTypeId);
		}
	},

	async upsertField(contentType, fieldId) {
		try {
			const ct = await getContentType(contentType.id);
			if (ct.fields.some((fld) => fld.id === fieldId)) {
				return contentType.editField(fieldId);
			}
			throw new Error(`Field ${fieldId} does not exist.`);
		} catch (error) {
			return contentType.createField(fieldId);
		}
	},
};
```

model.js:
```js
const {upsertContentType, upsertField} = require('./util');

module.exports = async function (migration) {
	console.log('get application');
	const application = await upsertContentType(migration, 'application');
	console.log('got application');
	application
		.name('Application')
		.displayField('displayName');
};
```

The event loop is exhausted and "got application" never prints. However if I short-circuit `upsertContentType` by returning a primitive like `'create'` or `{}`, the message does print and errors out appropriately. I'm using Node 8.6.0.

**Comment by zcei** [comment]
@jody-zeitler I just published `v0.3.0` which should fix your issue 🙂

**Comment by jody-zeitler** [comment]
Yes! The syntax is more awkward than I would like it to be, but it suits my needs.
